frontend-react/src/utils/DiffCompare/JsonDiffer.ts

```diff
@@ -0,0 +1,328 @@
+            {
+                start:
+                    jsonMap.pointers[each].key?.pos ||
+                    jsonMap.pointers[each].value.pos,
+                end: jsonMap.pointers[each].valueEnd.pos,
+            },
+        ],
+        []
+    );
+};
+
+/**
+ * Split out for unit testing
+ * The trick here is that we collect all the markup, then apply it
+ * working backwards so we don't screw with the offsets.
+ * e.g. "1, 2, 3, 4" vs "0, 1, 2, 5" if we highlight the  "0" on the right
+ * by inserting  "<mark>0</mark>, 1, 2, 5" then the offset for "5" has changed!
+ * but if we start at the end then the earlier offsets are the same.
+ * e.g. "0, 1, 2, <mark>5</mark>", the offset for "0" isn't changed when we get to it.
+ * @param jsonStr The json string to inject the markers into.
+ * @param markers Start/end pos for the various marks. Can be nested.
+ * @returns string with the <mark> tags inserted.
+ */
+const insertMarks = (jsonStr: string, markers: Marker[]): string => {
+    type MarkerInsert = {
+        pos: number;
+        mark: "<mark>" | "</mark>";
+    };
+    // turn into a single MarkerInsert[]. This enables easy back-to-front inserting into string
+    // we insert two entries per mark into the accumlator array
+    const inserts = markers.reduce(
+        (acc, each: Marker): MarkerInsert[] => [
+            ...acc,
+            { pos: each.start, mark: "<mark>" },
+            { pos: each.end, mark: "</mark>" },
+        ],
+        [] as MarkerInsert[]
+    );
+
+    // we reverse sort by pos. Work must be done from back to front
+    inserts.sort((a, b) => b.pos - a.pos);
+
+    // go through and inject <mark> or </mark> at each pos
+    return inserts.reduce(
+        (acc: string, each: MarkerInsert) =>
+            `${acc.slice(0, each.pos)}${each.mark}${acc.slice(each.pos)}`,
+        jsonStr
+    );
+};
+
+/**
+ * Diff compares two jsons and returns a map of the differences. You probably want to use jsonDifferMarkup
+ * @param leftData
+ * @param rightData
+ * @return JsonDiffResult Called by jsonDifferMarkup.
+ */
+const jsonDiffer = (
+    leftData: SourceMapResult,
+    rightData: SourceMapResult
+): JsonDiffResult => {
+    // diff the keys. If the key is different, then just consider the value of that key to be different.
+    const leftKeys = Object.keys(leftData.pointers);
+    const rightKeys = Object.keys(rightData.pointers);
+
+    // this is looking for diffs between the two lists.
+    let addedLeftKeys = leftKeys.filter(
+        (key) => key.length && !rightKeys.includes(key)
+    );
+    let addedRightKeys = rightKeys.filter(
+        (key) => key.length && !leftKeys.includes(key)
+    );
+
+    // now we want intersection (aka NOT changed and see if the values have changed).
+    let intersection = leftKeys.filter((key) => rightKeys.includes(key));
+
+    // for readability improvements only, pull out start/end values
+    const getStartEnd = (key: string, data: SourceMapResult) => {
```

Comment:
can these utility functions be split out of this parent function?

Reply:
They were originally just inlined. I split them out _just_ to improve readability.
`getStartEnd()` is only called just on the next line which is `getValue()`. 
And `getValue()` is only called on the next line `intersection.filter()`.